docs/Blog-Post.md

```diff
@@ -0,0 +1,525 @@
+
+y_pred = hax.dot(Feature, x, W)
+mse(y_pred, y)
+```
+
+The code is basically the same, but the presence of named axes mean that we don't accidentally broadcast `y` to the wrong shape.
+Instead, it works exactly as we intend.
+
+## Scale via Named Tensors
+
+We use named axes both to improve legibility and to enable scale: named axes are the basis of our
+[Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/) implementation as well as for tensor parallelism.
+FSDP can be added to a training loop with about 10 lines of code, enabling scale to at least 256 TPU cores (which is
+as many as we can get our hands on) and at least 65B parameters (which is way bigger than we have compute for).
+
+FSDP with Haliax basically amounts to telling Haliax which axes to shard, and specifying a different sharding for computation than for storage.
+A full tutorial is available [here](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz?usp=sharing), but here's a quick example:
+
+```diff
++# describe how we shard our parameters and our data
++# We store our parameters and optimizer states fully sharded along the embed axis
++param_mapping = {"embed": "data"}
++# During computation, we instead shard our data along the batch axis, and gather the parameters just-in-time
++data_mapping = {"batch": "data"}
+
++# tell Haliax to shard our model and optimizer states
++@hax.named_jit
+def init_model():
+-    return MyModel()
++    return hax.shard_with_axis_mapping(MyModel(), param_mapping)
+
+model = init_model()
+
+# initialize optimizer
+import optax
+optimizer = optax.adamw(1E-4, weight_decay=0.1)
+
++@hax.named_jit
+def init_optimizer(model):
+    opt_state = optimizer.init(model)
+-    return opt_state
++    return hax.shard_with_axis_mapping(opt_state, param_mapping)
+
+optimizer = init_optimizer(model)
+
++@hax.named_jit
+def train_step(model, opt_state, input_ids):
+  ... # elided for brevity
+
+  # ensure that intermediate states are sharded correctly
+-  loss, grads = grad_loss(model, input_ids)
++  with hax.axis_mapping(data_mapping):
++    loss, grads = grad_loss(model, input_ids)
+
+  ...
+  return loss, model, opt_state
+
+for data in data_iter:
++  data = hax.shard_with_axis_mapping(data, data_axis_mapping)
+  ...
+```
+
+Tensor parallelism can be added by simply changing the two axis mappings:
+
+```diff
+# Specify which axes we shard for tensor parallelism:
+# specifying "head" shards attention and "mlp" shards the feedforward
++tensor_parallel_mapping = {"head": "model", "mlp": "model"}
+# We store our parameters and optimizer states fully sharded along the embed axis
+-param_mapping = {"embed": "data"}
++param_mapping = {"embed": "data", **tensor_parallel_mapping}
+# During computation, we instead shard our data along the batch axis, and gather the parameters just-in-time
+-data_mapping = {"batch": "data"}
++data_mapping = {"batch": "data", **tensor_parallel_mapping}
+```
+
```

Comment:
This is cool!

Reply:
Thanks! I'm pretty sure I can get to pipeline parallelism this way too, but next release maybe.